# Credit Financial Risk Analysis

**Table of contents**

## Introduction
*   **Purpose:** The purpose of this project is to develop a Credit Risk Scoring Model that predicts the likelihood of company default using company profile, financial performance, credit history, and macroeconomic indicators. The model aims to support faster, safer, and more transparent lending decisions.

- **Dataset Overview:** <br>
The dataset contains records of **34,500 companies** with four main tables:

  - **company_info**:
    
  - **income statement**:
    
  - **balance sheet**:
    
  - **cash flow statement**:

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
company_info = pd.read_csv('/content/company_info.csv')
print(f"Company info(Total Row, Column): {company_info.shape}")

income = pd.read_csv('/content/income_statement.csv')
print(f"Income Statement(Total Row, Column): {income.shape}")

balance_sheet = pd.read_csv('/content/balance_sheet.csv')
print(f"Company info(Total Row, Column): {balance_sheet.shape}")

cash_flow = pd.read_csv('/content/cash_flow_statement.csv')
print(f"Company info(Total Row, Column): {cash_flow.shape}")

Company info(Total Row, Column): (34500, 4)
Income Statement(Total Row, Column): (154901, 13)
Company info(Total Row, Column): (154901, 23)
Company info(Total Row, Column): (154901, 19)


## Data Cleaning and validation
*   If one of the rows of a firm is invalid (which cannot be repaired/dropped), then the entire data of the firm concerned must also be dropped, because it can cause the year data to jump (there is an empty year gap).

*   **total invalid value = 32,61%**
*   **total invalid data recovered = 30,97%**
*   **total row = 154901**
*   **total row after cleaned = 150748 (97,3%)**

### **Company info**

Early data validation:
*   `firm_id`(PK) doesn't have a duplicate value
*   `sector` unique value : 'Jasa', 'Retail', 'Konstruksi', 'Manufaktur', 'F&B'
*   `region` unique value : 'Kalimantan', 'Jawa', 'Sumatera', 'Papua_Maluku', 'Sulawesi', 'Bali_NusaTenggara'
*   `start_year` range : 1990 - 2024
*   **There is no NULL value**



In [ ]:
company_info.head()

firm_id  sector      region  start_year
0  F000001    Jasa  Kalimantan        2021
1  F000002  Retail        Jawa        2000
2  F000003    Jasa    Sumatera        1995
3  F000004  Retail  Kalimantan        2019
4  F000005  Retail    Sumatera        2012

In [ ]:
company_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34500 entries, 0 to 34499
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   firm_id     34500 non-null  object
 1   sector      34500 non-null  object
 2   region      34500 non-null  object
 3   start_year  34500 non-null  int64 
dtypes: int64(1), object(3)
memory usage: 1.1+ MB


In [ ]:
company_info['firm_id'].duplicated().sum()

np.int64(0)

In [ ]:
company_info['sector'].unique()

array(['Jasa', 'Retail', 'Konstruksi', 'Manufaktur', 'F&B'], dtype=object)

In [ ]:
company_info['region'].unique()

array(['Kalimantan', 'Jawa', 'Sumatera', 'Papua_Maluku', 'Sulawesi',
       'Bali_NusaTenggara'], dtype=object)

In [ ]:
print(f'Range date: {company_info['start_year'].min()} - {company_info['start_year'].max()}')

Range date: 1990 - 2024


### **Income Statement**
*   **total invalid value = 19,12 + 1,09%**
*   **total invalid value recovered = 6.37%**


In [ ]:
income.head()

firm_id  year       revenue      cogs  gross_profit     opex   ebitda  \
0  F000001  2021  17410.340000   8702.19       8708.15  2833.64  5874.51   
1  F000001  2022 -10972.459916  15003.97      11215.11  4326.88  6888.23   
2  F000002  2018   2700.700000   2329.34        371.36   362.54     8.82   
3  F000002  2019   2944.410000   2220.01        724.40   353.91   370.49   
4  F000002  2020   1866.440000    818.04       1048.39   427.13   621.26   

   depreciation     ebit  interest_expense      ebt      tax  net_income  
0        609.00  5265.51            576.00  4689.52  1031.69     3657.82  
1        473.09  6415.14            702.87  5712.26  1256.70     4455.57  
2         59.04   -50.22             17.15   -67.37     0.00      -67.37  
3         62.15   308.34             39.02   269.32     1.35      267.97  
4         87.93   533.33             26.29   507.04     2.54      504.51

In [ ]:
income.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 154901 entries, 0 to 154900
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   firm_id           154901 non-null  object 
 1   year              154901 non-null  int64  
 2   revenue           152605 non-null  float64
 3   cogs              153351 non-null  float64
 4   gross_profit      154901 non-null  float64
 5   opex              153662 non-null  float64
 6   ebitda            154901 non-null  float64
 7   depreciation      154901 non-null  float64
 8   ebit              153043 non-null  float64
 9   interest_expense  154901 non-null  float64
 10  ebt               154901 non-null  float64
 11  tax               154901 non-null  float64
 12  net_income        154901 non-null  float64
dtypes: float64(11), int64(1), object(1)
memory usage: 15.4+ MB


In [ ]:
print(f'Date range: {income['year'].min()} - {income['year'].max()}')

Date range: 2018 - 2022


#### Null value detection (by row)
*   `revenue` percentage of null value = 1.48%
*   `cogs` percentage of null value = 1.00%
*   `opex` percentage of null value = 0.80%
*   `ebit` percentage of null value = 1.20%

**Total firms invalid data(all year each firm) = (29724)19,12%**

<br> Total recovered value (by row)
*   `revenue` percentage of recoverd = 1.45%
*   `cogs` percentage of recoverd = 1.00%
*   `opex` percentage of recoverd = 0.80%
*   `ebit` percentage of recoverd = 1.20%

<br>**total firms invalid data recovered(all year each firm) = 29557(19,08%)**


In [ ]:
income.isnull().sum()

firm_id                0
year                   0
revenue             2296
cogs                1550
gross_profit           0
opex                1239
ebitda                 0
depreciation           0
ebit                1858
interest_expense       0
ebt                    0
tax                    0
net_income             0
dtype: int64

In [ ]:
column_that_have_null = ['revenue', 'cogs', 'opex', 'ebit']
total_column = income.shape[0]

for column in column_that_have_null:
  null_value_each_column = income[column].isnull().sum()
  percentage = (null_value_each_column/total_column)*100
  print(f"{column} percentage of null value = {percentage:.2f}%")

revenue percentage of null value = 1.48%
cogs percentage of null value = 1.00%
opex percentage of null value = 0.80%
ebit percentage of null value = 1.20%


In [ ]:
total_null = income.isnull().sum().sum()
percentage_of_total = (total_null/income.shape[0])*100
print(f"Total percentage of null value = {percentage_of_total:.2f}%")

Total percentage of null value = 4.48%


In [ ]:
firms_with_null_income = income[income.isnull().any(axis=1)]
print("Firm IDs with null values in income DataFrame:")
display(firms_with_null_income['firm_id'].unique())

Firm IDs with null values in income DataFrame:


array(['F000005', 'F000008', 'F000020', ..., 'F034477', 'F034487',
       'F034489'], dtype=object)

In [ ]:
firms_with_null_income.shape

(6801, 13)

In [ ]:
firms_with_null_income_all_rows = income[income['firm_id'].isin(firms_with_null_income['firm_id'])]
firms_with_null_income_all_rows.shape

(29724, 13)

In [ ]:
income_valid = income.copy()

##### `revenue` null spotting and handling
*   Row that `cogs` and `gross_profit` not null is recovered with domain knowledge formula. (`cogs` + `gross_profit`)
*   Else, is being dropped. but we can find that in `invalid_fill_condition` dataframe

total missing value = 1.48%
<br>total recoverd = 1.45%

In [ ]:
income[income['revenue'].isnull()]

firm_id  year  revenue      cogs  gross_profit     opex   ebitda  \
213     F000047  2019      NaN   1133.58       1665.57   609.60  1055.97   
298     F000066  2021      NaN   1653.62        974.02   377.46   596.56   
323     F000072  2020      NaN    280.96        165.57    88.94    76.63   
350     F000077  2022      NaN  14426.48       5540.18  4381.02  1159.15   
625     F000141  2020      NaN    170.53        295.97    46.65   249.32   
...         ...   ...      ...       ...           ...      ...      ...   
154397  F034386  2022      NaN    579.76        307.74   142.44   165.29   
154449  F034399  2021      NaN    440.96        354.27    95.61   258.66   
154494  F034408  2022      NaN   2699.93        200.07   563.98  -363.92   
154671  F034448  2018      NaN    366.98        180.50   128.38    52.12   
154785  F034473  2022      NaN   2194.46        837.63   663.81   173.82   

        depreciation    ebit  interest_expense     ebt    tax  net_income  
213            73.25  982.72             43.20  939.51   4.70      934.82  
298           165.50  431.07             63.79  367.28   1.84      365.44  
323            18.39   58.24              3.52   54.72   0.27       54.44  
350           675.06  484.09            349.38  134.72  29.64      105.08  
625            13.60  235.72             13.56  222.16   1.11      221.05  
...              ...     ...               ...     ...    ...         ...  
154397         20.06  145.23              5.18  140.05   0.70      139.35  
154449         31.58  227.08              5.15  221.92   1.11      220.81  
154494        138.94 -502.86             44.99 -547.85   0.00     -547.85  
154671         12.85   39.27              1.34   37.93   0.19       37.74  
154785        151.95   21.87             64.86  -42.99   0.00      -42.99  

[2296 rows x 13 columns]

In [ ]:
# Observe the sample firm_id that have NaN revenue
income[income['firm_id'] == "F000047"]

firm_id  year  revenue     cogs  gross_profit    opex   ebitda  \
212  F000047  2018  2581.46  1929.35        652.12     NaN   123.84   
213  F000047  2019      NaN  1133.58       1665.57  609.60  1055.97   
214  F000047  2020  1708.72   622.30       1086.42  204.65   881.76   
215  F000047  2021  2064.32  1565.90        498.42  343.61   154.81   
216  F000047  2022  2574.52  1713.28        861.24  468.81   392.43   

     depreciation    ebit  interest_expense     ebt   tax  net_income  
212         71.88   51.97             27.98   23.99  0.12       23.87  
213         73.25  982.72             43.20  939.51  4.70      934.82  
214         61.78  819.98             41.72  778.26  3.89      774.36  
215        123.38   31.43             82.05  -50.62  0.00      -50.62  
216         67.09  325.34             74.45  250.89  1.25      249.63

firm_id in 2019 have NaN in revenue but have valid value in cogs and gross_profit. `gross_profit` formula is (`revenue` - `cogs`).

<br> because `gross_profit` in that case is have a valid value, we can fill that NaN with `gross_profit` + `cogs`. and we can observe is all of the NaN in `revenue` is the same case

In [ ]:
## How much Nan value of `revenue` that `cogs` and `gross_profit` value is not null?
filtered = income[income['revenue'].isnull() & income['cogs'].notnull() & income['gross_profit'].notnull()]
filtered

firm_id  year  revenue      cogs  gross_profit     opex   ebitda  \
213     F000047  2019      NaN   1133.58       1665.57   609.60  1055.97   
298     F000066  2021      NaN   1653.62        974.02   377.46   596.56   
323     F000072  2020      NaN    280.96        165.57    88.94    76.63   
350     F000077  2022      NaN  14426.48       5540.18  4381.02  1159.15   
625     F000141  2020      NaN    170.53        295.97    46.65   249.32   
...         ...   ...      ...       ...           ...      ...      ...   
154397  F034386  2022      NaN    579.76        307.74   142.44   165.29   
154449  F034399  2021      NaN    440.96        354.27    95.61   258.66   
154494  F034408  2022      NaN   2699.93        200.07   563.98  -363.92   
154671  F034448  2018      NaN    366.98        180.50   128.38    52.12   
154785  F034473  2022      NaN   2194.46        837.63   663.81   173.82   

        depreciation    ebit  interest_expense     ebt    tax  net_income  
213            73.25  982.72             43.20  939.51   4.70      934.82  
298           165.50  431.07             63.79  367.28   1.84      365.44  
323            18.39   58.24              3.52   54.72   0.27       54.44  
350           675.06  484.09            349.38  134.72  29.64      105.08  
625            13.60  235.72             13.56  222.16   1.11      221.05  
...              ...     ...               ...     ...    ...         ...  
154397         20.06  145.23              5.18  140.05   0.70      139.35  
154449         31.58  227.08              5.15  221.92   1.11      220.81  
154494        138.94 -502.86             44.99 -547.85   0.00     -547.85  
154671         12.85   39.27              1.34   37.93   0.19       37.74  
154785        151.95   21.87             64.86  -42.99   0.00      -42.99  

[2253 rows x 13 columns]

2253 of 2296 value can be imputed with calculation another variable.  

In [ ]:
invalid_fill_condition = income[income['revenue'].isnull() & (income['cogs'].isnull() | income['gross_profit'].isnull())]

invalid_fill_condition.shape

(43, 13)

The remaining 43 values are `cogs` and `revenue` both NaN

In [ ]:
invalid_firms_in_income_valid = income_valid[income_valid['firm_id'].isin(invalid_fill_condition['firm_id'])]
invalid_firms_in_income_valid.shape

(210, 13)

In [ ]:
## fill `revenue` NaN with `cogs` + `gross_profit`
income_valid['revenue_valid'] = income_valid['revenue'].fillna(income_valid['cogs'] + income_valid['gross_profit'])
income_valid.head()

firm_id  year       revenue      cogs  gross_profit     opex   ebitda  \
0  F000001  2021  17410.340000   8702.19       8708.15  2833.64  5874.51   
1  F000001  2022 -10972.459916  15003.97      11215.11  4326.88  6888.23   
2  F000002  2018   2700.700000   2329.34        371.36   362.54     8.82   
3  F000002  2019   2944.410000   2220.01        724.40   353.91   370.49   
4  F000002  2020   1866.440000    818.04       1048.39   427.13   621.26   

   depreciation     ebit  interest_expense      ebt      tax  net_income  \
0        609.00  5265.51            576.00  4689.52  1031.69     3657.82   
1        473.09  6415.14            702.87  5712.26  1256.70     4455.57   
2         59.04   -50.22             17.15   -67.37     0.00      -67.37   
3         62.15   308.34             39.02   269.32     1.35      267.97   
4         87.93   533.33             26.29   507.04     2.54      504.51   

   revenue_valid  
0   17410.340000  
1  -10972.459916  
2    2700.700000  
3    2944.410000  
4    1866.440000

In [ ]:
## remaining NaN in `revenue` varible because either `cogs` or `gross_profit` is NaN too.
income_valid['revenue_valid'].isnull().sum()

np.int64(43)

In [ ]:
## drop NaN value that `revenue` and `cogs` or `gross_profit` is NaN too
### but we can see dropped variable ini `invalid_fill_condition` dataframe
income_valid = income_valid.dropna(subset=['revenue_valid'])

In [ ]:
income_valid['revenue_valid'].isnull().sum()

np.int64(0)

In [ ]:
income_valid =  income_valid.drop(columns=['revenue'])
income_valid.head()

firm_id  year      cogs  gross_profit     opex   ebitda  depreciation  \
0  F000001  2021   8702.19       8708.15  2833.64  5874.51        609.00   
1  F000001  2022  15003.97      11215.11  4326.88  6888.23        473.09   
2  F000002  2018   2329.34        371.36   362.54     8.82         59.04   
3  F000002  2019   2220.01        724.40   353.91   370.49         62.15   
4  F000002  2020    818.04       1048.39   427.13   621.26         87.93   

      ebit  interest_expense      ebt      tax  net_income  revenue_valid  
0  5265.51            576.00  4689.52  1031.69     3657.82   17410.340000  
1  6415.14            702.87  5712.26  1256.70     4455.57  -10972.459916  
2   -50.22             17.15   -67.37     0.00      -67.37    2700.700000  
3   308.34             39.02   269.32     1.35      267.97    2944.410000  
4   533.33             26.29   507.04     2.54      504.51    1866.440000

##### `cogs` null spotting and handling
*    All of `cogs` Null Value can be recovered with domain knowledge calculation (revenue - gross_profit)

total missing value = 1%
<br>total recoverd = 1%

In [ ]:
income_valid['cogs'].isnull().sum()

np.int64(1507)

In [ ]:
## How much Nan value of `revenue` that `cogs` and `gross_profit` value is not null?
filtered = income_valid[income_valid['cogs'].isnull() & income_valid['revenue_valid'].notnull() & income_valid['gross_profit'].notnull()]
filtered.shape

(1507, 13)

all of the missing value in `cogs` is can be recovered with domain knowledge calculation(`revenue_valid` - `gross_profit`)

In [ ]:
income_valid['cogs_valid'] = income_valid['cogs'].fillna(income_valid['revenue_valid'] - income_valid['gross_profit'])
income_valid['cogs_valid'].isnull().sum()

np.int64(0)

In [ ]:
income_valid = income_valid.drop(columns=['cogs'])
income_valid.head()

firm_id  year  gross_profit     opex   ebitda  depreciation     ebit  \
0  F000001  2021       8708.15  2833.64  5874.51        609.00  5265.51   
1  F000001  2022      11215.11  4326.88  6888.23        473.09  6415.14   
2  F000002  2018        371.36   362.54     8.82         59.04   -50.22   
3  F000002  2019        724.40   353.91   370.49         62.15   308.34   
4  F000002  2020       1048.39   427.13   621.26         87.93   533.33   

   interest_expense      ebt      tax  net_income  revenue_valid  cogs_valid  
0            576.00  4689.52  1031.69     3657.82   17410.340000     8702.19  
1            702.87  5712.26  1256.70     4455.57  -10972.459916    15003.97  
2             17.15   -67.37     0.00      -67.37    2700.700000     2329.34  
3             39.02   269.32     1.35      267.97    2944.410000     2220.01  
4             26.29   507.04     2.54      504.51    1866.440000      818.04

##### `Opex` null spotting & handling
*    All of `opex` Null Value can be recovered with domain knowledge calculation (`gross_profit` - `ebitda`)
total missing value = 0,80%
<br>total recoverd = 0,80%

In [ ]:
income_valid['opex'].isnull().sum()

np.int64(1239)

In [ ]:
income_valid[income_valid['opex'].isnull()]

firm_id  year  gross_profit  opex   ebitda  depreciation     ebit  \
30      F000008  2021        184.89   NaN   116.63         20.45    96.18   
212     F000047  2018        652.12   NaN   123.84         71.88    51.97   
315     F000070  2018         94.56   NaN     5.68         10.98    -5.30   
669     F000151  2021        160.19   NaN   -93.30         62.70  -156.00   
677     F000153  2019       2774.10   NaN  1936.79         77.85  1858.94   
...         ...   ...           ...   ...      ...           ...      ...   
153988  F034294  2018        177.31   NaN    66.16         12.72    53.45   
154035  F034304  2019         90.03   NaN    30.09          8.68    21.40   
154136  F034326  2021        762.98   NaN  -147.67        221.78  -369.46   
154253  F034352  2022        207.86   NaN   117.67         14.31   103.36   
154582  F034428  2020        519.13   NaN -1122.32        426.08 -1548.40   

        interest_expense      ebt   tax  net_income  revenue_valid  cogs_valid  
30                  4.02    92.16  0.46       91.70         352.40      167.52  
212                27.98    23.99  0.12       23.87        2581.46     1929.35  
315                 5.08   -10.38  0.00      -10.38         355.83      261.27  
669                26.30  -182.30  0.00     -182.30        1389.09     1228.90  
677                90.96  1767.98  8.84     1759.14        4355.72     1581.62  
...                  ...      ...   ...         ...            ...         ...  
153988              7.30    46.14  0.23       45.91         562.76      385.45  
154035              2.76    18.64  0.09       18.55         402.94      312.90  
154136            167.75  -537.21  0.00     -537.21        4096.56     3333.58  
154253              4.04    99.32  0.50       98.82         390.10      182.23  
154582            545.25 -2093.65  0.00    -2093.65        9503.01     8983.88  

[1239 rows x 13 columns]

In [ ]:
filtered = income_valid[income_valid['opex'].isnull() & income_valid['gross_profit'].notnull() & income_valid['ebitda'].notnull()]
filtered.shape

(1239, 13)

all of the `opex` null value can be recovered with domain knowledge(`gross_profit` - `ebitda`)

In [ ]:
income_valid['opex_valid'] = income_valid['opex'].fillna(income_valid['gross_profit'] - income_valid['ebitda'])
income_valid = income_valid.drop(columns=['opex'])
income_valid.head()

firm_id  year  gross_profit   ebitda  depreciation     ebit  \
0  F000001  2021       8708.15  5874.51        609.00  5265.51   
1  F000001  2022      11215.11  6888.23        473.09  6415.14   
2  F000002  2018        371.36     8.82         59.04   -50.22   
3  F000002  2019        724.40   370.49         62.15   308.34   
4  F000002  2020       1048.39   621.26         87.93   533.33   

   interest_expense      ebt      tax  net_income  revenue_valid  cogs_valid  \
0            576.00  4689.52  1031.69     3657.82   17410.340000     8702.19   
1            702.87  5712.26  1256.70     4455.57  -10972.459916    15003.97   
2             17.15   -67.37     0.00      -67.37    2700.700000     2329.34   
3             39.02   269.32     1.35      267.97    2944.410000     2220.01   
4             26.29   507.04     2.54      504.51    1866.440000      818.04   

   opex_valid  
0     2833.64  
1     4326.88  
2      362.54  
3      353.91  
4      427.13

In [ ]:
income_valid['opex_valid'].isnull().sum()

np.int64(0)

##### `ebit` null spotting & handling
*    All of `ebit` Null Value can be recovered with domain knowledge calculation (ebitda - depreciation)
total missing value = 1,2%
<br>total recoverd = 1,2%

In [ ]:
income_valid['ebit'].isnull().sum()

np.int64(1858)

In [ ]:
filtered = income_valid[income_valid['ebit'].isnull() & income_valid['ebitda'].notnull() & income_valid['depreciation'].notnull()]
filtered.shape

(1858, 13)

In [ ]:
income_valid.head()

firm_id  year  gross_profit   ebitda  depreciation     ebit  \
0  F000001  2021       8708.15  5874.51        609.00  5265.51   
1  F000001  2022      11215.11  6888.23        473.09  6415.14   
2  F000002  2018        371.36     8.82         59.04   -50.22   
3  F000002  2019        724.40   370.49         62.15   308.34   
4  F000002  2020       1048.39   621.26         87.93   533.33   

   interest_expense      ebt      tax  net_income  revenue_valid  cogs_valid  \
0            576.00  4689.52  1031.69     3657.82   17410.340000     8702.19   
1            702.87  5712.26  1256.70     4455.57  -10972.459916    15003.97   
2             17.15   -67.37     0.00      -67.37    2700.700000     2329.34   
3             39.02   269.32     1.35      267.97    2944.410000     2220.01   
4             26.29   507.04     2.54      504.51    1866.440000      818.04   

   opex_valid  
0     2833.64  
1     4326.88  
2      362.54  
3      353.91  
4      427.13

In [ ]:
income_valid['ebit_valid'] = income_valid['ebit'].fillna(income_valid['ebitda'] - income_valid['depreciation'])
income_valid = income_valid.drop(columns=['ebit'])
income_valid.head()

firm_id  year  gross_profit   ebitda  depreciation  interest_expense  \
0  F000001  2021       8708.15  5874.51        609.00            576.00   
1  F000001  2022      11215.11  6888.23        473.09            702.87   
2  F000002  2018        371.36     8.82         59.04             17.15   
3  F000002  2019        724.40   370.49         62.15             39.02   
4  F000002  2020       1048.39   621.26         87.93             26.29   

       ebt      tax  net_income  revenue_valid  cogs_valid  opex_valid  \
0  4689.52  1031.69     3657.82   17410.340000     8702.19     2833.64   
1  5712.26  1256.70     4455.57  -10972.459916    15003.97     4326.88   
2   -67.37     0.00      -67.37    2700.700000     2329.34      362.54   
3   269.32     1.35      267.97    2944.410000     2220.01      353.91   
4   507.04     2.54      504.51    1866.440000      818.04      427.13   

   ebit_valid  
0     5265.51  
1     6415.14  
2      -50.22  
3      308.34  
4      533.33

In [ ]:
income_valid.isnull().sum()

firm_id             0
year                0
gross_profit        0
ebitda              0
depreciation        0
interest_expense    0
ebt                 0
tax                 0
net_income          0
revenue_valid       0
cogs_valid          0
opex_valid          0
ebit_valid          0
dtype: int64

#### Negative value detection
Variable that impossible/not make sense have negative value
*    `revenue` = 450 row (0.29%)
*    `cogs` = 0 row
*    `opex` = 0 row

**Total firms invalid data(all year each firm) = 2136 row (1.09%)**
<br>**total firms invalid data recovered(all year each firm) = 1.09%**

In [ ]:
negative_revenue = income_valid[income_valid['revenue_valid'] < 0 ]
print(f'total negative revenue: {negative_revenue.shape[0]}')
print(f'percentage negative revenue: {(negative_revenue.shape[0]/income_valid.shape[0])*100}')

total negative revenue: 450
percentage negative revenue: 0.2905887974789807


In [ ]:
firms_with_negative_revenue_in_income_valid = income_valid[income_valid['firm_id'].isin(negative_revenue['firm_id'])]
firms_with_negative_revenue_in_income_valid.shape

(2138, 13)

In [ ]:
income_valid[income_valid['revenue_valid']<0]

firm_id  year  gross_profit   ebitda  depreciation  interest_expense  \
1       F000001  2022      11215.11  6888.23        473.09            702.87   
221     F000049  2021         90.66    -0.70         19.47              4.78   
227     F000050  2022       4544.11  3082.58        397.36            243.85   
581     F000132  2021        349.87   235.59         11.99             12.64   
740     F000166  2020       1096.27  -289.70        151.53            269.92   
...         ...   ...           ...      ...           ...               ...   
154059  F034309  2019        314.81   233.87         15.11              3.34   
154123  F034323  2022        203.96    75.21         15.24              5.68   
154236  F034349  2020        814.95   492.70        127.60             67.60   
154246  F034351  2020        447.18   229.47         40.90             74.92   
154687  F034452  2018         27.76   -25.65          8.69              3.80   

            ebt      tax  net_income  revenue_valid  cogs_valid  opex_valid  \
1       5712.26  1256.70     4455.57  -10972.459916    15003.97     4326.88   
221      -24.94     0.00      -24.94    -183.335768      421.68       91.35   
227     2441.38   537.10     1904.27   -2498.713423    10071.15     1461.53   
581      210.96     1.05      209.90    -198.606635      225.72      114.29   
740     -711.14     0.00     -711.14   -1409.802630     6915.01     1385.97   
...         ...      ...         ...            ...         ...         ...   
154059   215.43     1.08      214.35     -78.385849      180.71       80.94   
154123    54.29     0.27       54.02    -240.791993      237.00      128.75   
154236   297.49     1.49      296.01    -810.308906     1287.10      322.25   
154246   113.65     0.57      113.08   -1307.759343     1098.18      217.70   
154687   -38.14     0.00      -38.14    -262.845259      337.08       53.41   

        ebit_valid  
1          6415.14  
221         -20.17  
227        2685.23  
581         223.59  
740        -441.23  
...            ...  
154059      218.77  
154123       59.97  
154236      365.10  
154246      188.57  
154687      -34.34  

[450 rows x 13 columns]

In [ ]:
income_valid['revenue_valid'] = np.where(
    income_valid['revenue_valid'] < 0,
    np.nan,
    income_valid['revenue_valid']
)

income_valid[income_valid['revenue_valid']<0]

Empty DataFrame
Columns: [firm_id, year, gross_profit, ebitda, depreciation, interest_expense, ebt, tax, net_income, revenue_valid, cogs_valid, opex_valid, ebit_valid]
Index: []

In [ ]:
income_valid['revenue_valid'] = income_valid['revenue_valid'].fillna(income_valid['cogs_valid'] + income_valid['gross_profit'])
income_valid.isnull().sum()

firm_id             0
year                0
gross_profit        0
ebitda              0
depreciation        0
interest_expense    0
ebt                 0
tax                 0
net_income          0
revenue_valid       0
cogs_valid          0
opex_valid          0
ebit_valid          0
dtype: int64

In [ ]:
invalid_fill_condition2 = income_valid[income_valid['revenue_valid']<0]
invalid_fill_condition2

firm_id  year  gross_profit  ebitda  depreciation  interest_expense  \
81766   F018176  2020        328.77  -80.81        183.20             51.62   
112555  F025025  2019       1033.19  541.86        102.99             43.28   

           ebt   tax  net_income  revenue_valid   cogs_valid  opex_valid  \
81766  -315.63  0.00     -315.63   -1976.324763 -2305.094763      409.57   
112555  395.59  1.98      393.61    -999.439726 -2032.629726      491.33   

        ebit_valid  
81766      -264.01  
112555      438.87

It turns out that there is still data whose revenues are negative

In [ ]:
income[income['firm_id'] == "F018176"]

firm_id  year      revenue     cogs  gross_profit    opex  ebitda  \
81764  F018176  2018  2946.560000  2399.92        546.65  493.33   53.32   
81765  F018176  2019  2917.700000  2221.01        696.69  485.32  211.36   
81766  F018176  2020 -1976.324763      NaN        328.77  409.57  -80.81   
81767  F018176  2021  3060.690000  1765.53       1295.17  662.19  632.98   
81768  F018176  2022  3790.240000  2593.28       1196.95  574.06  622.89   

       depreciation    ebit  interest_expense     ebt   tax  net_income  
81764        103.70  -50.38             39.61  -89.98  0.00      -89.98  
81765         80.41  130.95             62.50   68.46  0.34       68.11  
81766        183.20 -264.01             51.62 -315.63  0.00     -315.63  
81767         93.00  539.97             75.79  464.18  2.32      461.86  
81768         90.86  532.03             60.08  471.95  2.36      469.59

After observing that is because the original data from the table has negative revenue and NaN cogs. which is the same as NaN. so because it is crucial data, we drop this data row.

We decide to save dataframe that have invalid value on cogs and revenue to be followed up by the officer by requesting more data from the creditor with lister firm_id.

In [ ]:
invalid_fill_condition.to_csv('/content/invalid_data/firms_invalid_revenue_cogs.csv', index=False)

In [ ]:
invalid_fill_condition2.to_csv('/content/invalid_data/firms_invalid_revenue_cogs2.csv', index=False)

In [ ]:
income_valid = income_valid.drop(index=invalid_fill_condition2.index)

In [ ]:
invalid_fill_condition2 = income_valid[income_valid['revenue_valid']<0]
invalid_fill_condition2

Empty DataFrame
Columns: [firm_id, year, gross_profit, ebitda, depreciation, interest_expense, ebt, tax, net_income, revenue_valid, cogs_valid, opex_valid, ebit_valid]
Index: []

In [ ]:
## spot negative cogs
negative_cogs = income[income['cogs'] < 0 ]
print(f'total negative cogs: {negative_cogs.shape[0]}')
print(f'percentage negative cogs: {(negative_cogs.shape[0]/income.shape[0])*100}')

total negative cogs: 0
percentage negative cogs: 0.0


In [ ]:
## spot negative opex
negative_opex = income[income['opex'] < 0 ]
print(f'total negative opex: {negative_opex.shape[0]}')
print(f'percentage negative opex: {(negative_opex.shape[0]/income.shape[0])*100}')

total negative opex: 0
percentage negative opex: 0.0


#### Impossible value detection
*    Value `cogs` that bigger than revenue = 0,46%
*    gross profit !=  revenue - cogs = 1,18%
total impossible value = 1,64%
<br>total impossible value recovered = 1,64%

##### Cogs > revenue
*   total value = 754
*   percentage = 0.46%

**Total firms invalid data(all year each firm) = 3598(2,3%)**
**<br>Total firms invalid data recovered(all year each firm) = 2.3%**

In [ ]:
# Spot cogs > revenue
impossible_value = income_valid['cogs_valid'] > income_valid['revenue_valid']
total = impossible_value.sum()
print(f'total impossible value: {total}')
print(f'percentage of impossible value: {(total/income_valid.shape[0])*100}%')

total impossible value: 754
percentage of impossible value: 0.48690396239086636%


In [ ]:
impossible_value_df = income_valid[impossible_value]

firms_invalid = income_valid[income_valid['firm_id'].isin(impossible_value_df['firm_id'])]
firms_invalid.shape

(3598, 13)

In [ ]:
impossible_value_df = income_valid[impossible_value]
impossible_value_df

firm_id  year  gross_profit   ebitda  depreciation  interest_expense  \
110     F000026  2020         65.15   -57.11         19.04              7.59   
1453    F000326  2022        285.92   167.74         17.02             16.29   
1632    F000367  2020       9472.97  7288.10        698.49            454.10   
2297    F000516  2020         24.31   -85.26         15.21             14.05   
2442    F000552  2018       2399.48   -56.80        471.72            235.83   
...         ...   ...           ...      ...           ...               ...   
153425  F034162  2019       2929.82  2058.91        158.94            296.52   
153655  F034217  2018         97.80   -23.09         13.06              5.88   
153744  F034236  2021        189.42   100.17         10.11              4.23   
153751  F034238  2018         67.97    15.36         12.20              4.38   
154325  F034370  2020        794.63   -47.00        268.86            331.23   

            ebt      tax  net_income  revenue_valid    cogs_valid  opex_valid  \
110      -83.74     0.00      -83.74         501.72    666.039595      122.27   
1453     134.43     0.67      133.76         526.65    609.857105      118.18   
1632    6135.51  1349.81     4785.70       21848.70  24360.657317     2184.87   
2297    -114.51     0.00     -114.51         436.84    616.267106      109.56   
2442    -764.35     0.00     -764.35       16488.93  20461.489754     2456.28   
...         ...      ...         ...            ...           ...         ...   
153425  1603.45   352.76     1250.69        8029.55   9057.676535      870.92   
153655   -42.02     0.00      -42.02         734.81    908.928860      120.89   
153744    85.82     0.43       85.40         352.71    434.918724       89.25   
153751    -1.23     0.00       -1.23         487.48    636.389937       52.61   
154325  -647.08     0.00     -647.08        8416.21  11388.944343      841.62   

        ebit_valid  
110         -76.15  
1453        150.72  
1632       6589.61  
2297       -100.46  
2442       -528.52  
...            ...  
153425     1899.97  
153655      -36.15  
153744       90.06  
153751        3.16  
154325     -315.85  

[754 rows x 13 columns]

With assumtion that `opex_valid` is correct, the value of ebitda is being tested. is `ebitda` = `gross_profit` - `opex_valid` and give 0,01 tolerance if there any different result because of floating point.

In [ ]:
impossible_value_df['ebitda_test'] = impossible_value_df['gross_profit'] - impossible_value_df['opex_valid']

tolerance = 1e-2

impossible_value_df['is_ebitda_valid'] = np.isclose(
    impossible_value_df['ebitda_test'],
    impossible_value_df['ebitda'],
    atol=tolerance
)

invalid_ebitda = impossible_value_df[~impossible_value_df['is_ebitda_valid']]
invalid_ebitda

/tmp/ipython-input-3904914157.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  impossible_value_df['ebitda_test'] = impossible_value_df['gross_profit'] - impossible_value_df['opex_valid']
/tmp/ipython-input-3904914157.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  impossible_value_df['is_ebitda_valid'] = np.isclose(


Empty DataFrame
Columns: [firm_id, year, gross_profit, ebitda, depreciation, interest_expense, ebt, tax, net_income, revenue_valid, cogs_valid, opex_valid, ebit_valid, ebitda_test, is_ebitda_valid]
Index: []

In [ ]:
impossible_value_df['is_ebitda_valid'].unique()

array([ True])

unique value of `is_ebitda_valid` is all true. it indicates that we can trust `gross_profit` value. after that, we just cant change `cogs` value that > `revenue` with doing `gross_profit` calculation

In [ ]:
# Make cogs > revenue to NaN
income_valid.loc[income_valid['cogs_valid'] > income_valid['revenue_valid'], 'cogs_valid'] = np.nan
income_valid.isnull().sum()

firm_id               0
year                  0
gross_profit          0
ebitda                0
depreciation          0
interest_expense      0
ebt                   0
tax                   0
net_income            0
revenue_valid         0
cogs_valid          754
opex_valid            0
ebit_valid            0
dtype: int64

In [ ]:
income_valid['cogs_valid'].fillna(income_valid['revenue_valid'] - income_valid['gross_profit'], inplace=True)
income_valid.isnull().sum()

/tmp/ipython-input-3781483811.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  income_valid['cogs_valid'].fillna(income_valid['revenue_valid'] - income_valid['gross_profit'], inplace=True)


firm_id             0
year                0
gross_profit        0
ebitda              0
depreciation        0
interest_expense    0
ebt                 0
tax                 0
net_income          0
revenue_valid       0
cogs_valid          0
opex_valid          0
ebit_valid          0
dtype: int64

In [ ]:
impossible_value = income_valid['cogs_valid'] > income_valid['revenue_valid']
total = impossible_value.sum()
print(f'total impossible value: {total}')
print(f'percentage of impossible value: {(total/income_valid.shape[0])*100}%')

total impossible value: 0
percentage of impossible value: 0.0%


##### gross profit != revenue - cogs
*    total value = 1841
*    percentage = 1,18%

**total recovered = 1,18%**
<br>**total invalid firm data = 8459(5,4%)**

In [ ]:
mask = ~np.isclose(
    income_valid['gross_profit'],
    income_valid['revenue_valid'] - income_valid['cogs_valid'],
    atol=1e-2  # toleransi 0.01
)

impossible_value_df2 = income_valid[mask]
impossible_value_df2

firm_id  year  gross_profit   ebitda  depreciation  interest_expense  \
143     F000033  2018        112.33    16.24         13.67              3.10   
345     F000076  2022        273.96   106.57         55.85             26.33   
367     F000081  2022        171.88   -71.37         21.67              6.11   
446     F000101  2018       1173.16   654.10         85.69             38.15   
573     F000130  2021        364.94   264.07         15.56              6.26   
...         ...   ...           ...      ...           ...               ...   
154547  F034420  2021        236.72    50.21         34.50              4.86   
154612  F034434  2021       1837.74  1237.00        111.63             92.89   
154658  F034444  2021        747.03   629.57         32.52             42.15   
154715  F034457  2022       1201.83   320.39         90.42             73.53   
154790  F034474  2022         52.38   -62.34         21.33              3.27   

            ebt   tax  net_income  revenue_valid  cogs_valid  opex_valid  \
143       -0.53  0.00       -0.53   57215.970594      487.78       96.09   
345       24.38  0.12       24.26   72832.111201     1399.97      167.39   
367      -99.16  0.00      -99.16   62600.852348      789.25      243.25   
446      530.25  2.65      527.60   70611.819926     1426.06      519.07   
573      242.25  1.21      241.04   97669.280087      248.95      100.86   
...         ...   ...         ...            ...         ...         ...   
154547    10.85  0.05       10.80   97021.977312      763.28      186.51   
154612  1032.48  5.16     1027.32   61583.259675     1742.82      600.74   
154658   554.90  2.77      552.13   95019.450842      427.58      117.46   
154715   156.44  0.78      155.66   55853.512075     3013.70      881.44   
154790   -86.94  0.00      -86.94   96287.561708      421.60      114.73   

        ebit_valid  
143           2.58  
345          50.71  
367         -93.05  
446         568.40  
573         248.51  
...            ...  
154547       15.71  
154612     1125.37  
154658      597.05  
154715      229.97  
154790      -83.67  

[1841 rows x 13 columns]

cyou can see that the list of dataframe values has a high `revenue` of >50M. which violates the maximum value of MSME revenue according to state law. maybe this is a revenue input error.

In [ ]:
firms_invalid2 = income_valid[income_valid['firm_id'].isin(impossible_value_df2['firm_id'])]
firms_invalid2.shape

(8459, 13)

In [ ]:
impossible_value_df2[impossible_value_df2['revenue_valid'] > 50000]

firm_id  year  gross_profit   ebitda  depreciation  interest_expense  \
143     F000033  2018        112.33    16.24         13.67              3.10   
345     F000076  2022        273.96   106.57         55.85             26.33   
367     F000081  2022        171.88   -71.37         21.67              6.11   
446     F000101  2018       1173.16   654.10         85.69             38.15   
573     F000130  2021        364.94   264.07         15.56              6.26   
...         ...   ...           ...      ...           ...               ...   
154547  F034420  2021        236.72    50.21         34.50              4.86   
154612  F034434  2021       1837.74  1237.00        111.63             92.89   
154658  F034444  2021        747.03   629.57         32.52             42.15   
154715  F034457  2022       1201.83   320.39         90.42             73.53   
154790  F034474  2022         52.38   -62.34         21.33              3.27   

            ebt   tax  net_income  revenue_valid  cogs_valid  opex_valid  \
143       -0.53  0.00       -0.53   57215.970594      487.78       96.09   
345       24.38  0.12       24.26   72832.111201     1399.97      167.39   
367      -99.16  0.00      -99.16   62600.852348      789.25      243.25   
446      530.25  2.65      527.60   70611.819926     1426.06      519.07   
573      242.25  1.21      241.04   97669.280087      248.95      100.86   
...         ...   ...         ...            ...         ...         ...   
154547    10.85  0.05       10.80   97021.977312      763.28      186.51   
154612  1032.48  5.16     1027.32   61583.259675     1742.82      600.74   
154658   554.90  2.77      552.13   95019.450842      427.58      117.46   
154715   156.44  0.78      155.66   55853.512075     3013.70      881.44   
154790   -86.94  0.00      -86.94   96287.561708      421.60      114.73   

        ebit_valid  
143           2.58  
345          50.71  
367         -93.05  
446         568.40  
573         248.51  
...            ...  
154547       15.71  
154612     1125.37  
154658      597.05  
154715      229.97  
154790      -83.67  

[1841 rows x 13 columns]

and sure enough, the `gross_profit` values are not equal to `revenue` - `cogs`, the `revenue` values are all >50M. therefore I will consider this invalid by replacing it with a calculation using `gross_profit` and `cogs` as the benchmark.

In [ ]:
income_valid.loc[mask, 'revenue_valid'] = np.nan
income_valid.isnull().sum()

firm_id                0
year                   0
gross_profit           0
ebitda                 0
depreciation           0
interest_expense       0
ebt                    0
tax                    0
net_income             0
revenue_valid       1841
cogs_valid             0
opex_valid             0
ebit_valid             0
dtype: int64

In [ ]:
income_valid['revenue_valid'].fillna(income_valid['gross_profit'] + income_valid['cogs_valid'], inplace=True)
income_valid.isnull().sum()

/tmp/ipython-input-103329017.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  income_valid['revenue_valid'].fillna(income_valid['gross_profit'] + income_valid['cogs_valid'], inplace=True)


firm_id             0
year                0
gross_profit        0
ebitda              0
depreciation        0
interest_expense    0
ebt                 0
tax                 0
net_income          0
revenue_valid       0
cogs_valid          0
opex_valid          0
ebit_valid          0
dtype: int64

In [ ]:
income_valid.head()

firm_id  year  gross_profit   ebitda  depreciation  interest_expense  \
0  F000001  2021       8708.15  5874.51        609.00            576.00   
1  F000001  2022      11215.11  6888.23        473.09            702.87   
2  F000002  2018        371.36     8.82         59.04             17.15   
3  F000002  2019        724.40   370.49         62.15             39.02   
4  F000002  2020       1048.39   621.26         87.93             26.29   

       ebt      tax  net_income  revenue_valid  cogs_valid  opex_valid  \
0  4689.52  1031.69     3657.82       17410.34     8702.19     2833.64   
1  5712.26  1256.70     4455.57       26219.08    15003.97     4326.88   
2   -67.37     0.00      -67.37        2700.70     2329.34      362.54   
3   269.32     1.35      267.97        2944.41     2220.01      353.91   
4   507.04     2.54      504.51        1866.44      818.04      427.13   

   ebit_valid  
0     5265.51  
1     6415.14  
2      -50.22  
3      308.34  
4      533.33

### **Balance Sheet Overview**
*   There is no NULL value
*   total negative value = 1144(0,7%)

**total invalid value recovered** = **472 row (0.3%)**
<br>**total row after cleaning = 153847**

In [ ]:
pd.set_option('display.max_columns', None)

In [ ]:
balance_sheet.head()

firm_id  year      cash  receivables  inventory  other_current_assets  \
0  F000001  2021  16122.08      1723.84       0.00                181.27   
1  F000001  2022  16625.47      2699.45       0.00                313.41   
2  F000002  2018    948.86       260.10     272.04                 40.75   
3  F000002  2019   1357.06        57.56     184.94                 82.95   
4  F000002  2020   1719.01        89.09     118.03                 26.36   

   total_current_assets  ppe_gross  accum_depreciation  ppe_net  \
0              18027.19    4300.20              609.00  3691.20   
1              19638.33    4563.02             1082.09  3480.92   
2               1521.75     420.47               59.04   361.43   
3               1682.51     483.79              121.19   362.59   
4               1952.48     565.37              209.12   356.25   

   other_noncurrent_assets  total_assets  payables  other_current_liabilities  \
0                   134.54      21852.92    963.81                     444.00   
1                   253.69      23372.94    936.89                     933.45   
2                    48.61       1931.79    193.44                      79.62   
3                    52.93       2098.03    246.67                     118.26   
4                    20.64       2329.37     68.52                      54.70   

   current_debt  total_current_liabilities  long_term_debt  total_liabilities  \
0       2659.88                    4067.70         4939.78            9007.48   
1        706.58                    2576.92         4710.53            7287.45   
2         92.38                     365.45          171.57             537.02   
3         24.44                     389.37          162.92             552.30   
4         24.44                     147.66          162.92             310.59   

   equity_begin  dividends  equity_injection  equity_end  \
0      10159.91     972.29              0.00    12845.44   
1      12845.44    1215.52              0.00    16085.49   
2       1392.52       0.00             69.63     1394.77   
3       1394.77     117.01              0.00     1545.73   
4       1545.73      31.45              0.00     2018.78   

   total_liabilities_and_equity  
0                      21852.92  
1                      23372.94  
2                       1931.79  
3                       2098.03  
4                       2329.37

In [ ]:
balance_sheet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 154901 entries, 0 to 154900
Data columns (total 23 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   firm_id                       154901 non-null  object 
 1   year                          154901 non-null  int64  
 2   cash                          154901 non-null  float64
 3   receivables                   154901 non-null  float64
 4   inventory                     154901 non-null  float64
 5   other_current_assets          154901 non-null  float64
 6   total_current_assets          154901 non-null  float64
 7   ppe_gross                     154901 non-null  float64
 8   accum_depreciation            154901 non-null  float64
 9   ppe_net                       154901 non-null  float64
 10  other_noncurrent_assets       154901 non-null  float64
 11  total_assets                  154901 non-null  float64
 12  payables                      154901 non-nul

#### Filtering invalid firm from income_statement

In [ ]:
filter1 = pd.read_csv('/content/invalid_data/firms_invalid_revenue_cogs.csv')
filter1['firm_id'].unique()

array(['F000212', 'F000217', 'F002732', 'F002876', 'F004303', 'F006122',
       'F006804', 'F007461', 'F008189', 'F008781', 'F010425', 'F010863',
       'F011925', 'F012700', 'F013947', 'F014531', 'F015133', 'F015944',
       'F016008', 'F016027', 'F016309', 'F017404', 'F018015', 'F018220',
       'F018257', 'F019407', 'F021619', 'F021985', 'F022771', 'F024114',
       'F024294', 'F025963', 'F026459', 'F027102', 'F027143', 'F027203',
       'F029762', 'F031930', 'F032049', 'F033046', 'F033091', 'F033329',
       'F033397'], dtype=object)

In [ ]:
filter2 = pd.read_csv('/content/invalid_data/firms_invalid_revenue_cogs2.csv')
filter2['firm_id'].unique()

array(['F018176', 'F025025'], dtype=object)

In [ ]:
invalid_firm_ids = pd.concat([filter1['firm_id'], filter2['firm_id']]).unique()
print(f'Total number of invalid firm IDs: {len(invalid_firm_ids)}')
print(f'Example invalid firm IDs: {invalid_firm_ids[:5]}')

Total number of invalid firm IDs: 45
Example invalid firm IDs: ['F000212' 'F000217' 'F002732' 'F002876' 'F004303']


In [ ]:
balance_sheet_filter1 = balance_sheet[~balance_sheet['firm_id'].isin(filter1['firm_id'].unique())]
balance_sheet_filter1[balance_sheet_filter1['firm_id'] == 'F000212']

Empty DataFrame
Columns: [firm_id, year, cash, receivables, inventory, other_current_assets, total_current_assets, ppe_gross, accum_depreciation, ppe_net, other_noncurrent_assets, total_assets, payables, other_current_liabilities, current_debt, total_current_liabilities, long_term_debt, total_liabilities, equity_begin, dividends, equity_injection, equity_end, total_liabilities_and_equity]
Index: []

In [ ]:
balance_sheet_filter2 = balance_sheet_filter1[~balance_sheet_filter1['firm_id'].isin(filter2['firm_id'].unique())]
balance_sheet_filter2[balance_sheet_filter2['firm_id'] == 'F018176']

Empty DataFrame
Columns: [firm_id, year, cash, receivables, inventory, other_current_assets, total_current_assets, ppe_gross, accum_depreciation, ppe_net, other_noncurrent_assets, total_assets, payables, other_current_liabilities, current_debt, total_current_liabilities, long_term_debt, total_liabilities, equity_begin, dividends, equity_injection, equity_end, total_liabilities_and_equity]
Index: []

In [ ]:
balance_sheet_filter2.shape

(154681, 23)

In [ ]:
balance_sheet_filter2.isnull().sum()

firm_id                         0
year                            0
cash                            0
receivables                     0
inventory                       0
other_current_assets            0
total_current_assets            0
ppe_gross                       0
accum_depreciation              0
ppe_net                         0
other_noncurrent_assets         0
total_assets                    0
payables                        0
other_current_liabilities       0
current_debt                    0
total_current_liabilities       0
long_term_debt                  0
total_liabilities               0
equity_begin                    0
dividends                       0
equity_injection                0
equity_end                      0
total_liabilities_and_equity    0
dtype: int64

In [ ]:
balance_sheet_filter2.describe()

year           cash    receivables      inventory  \
count  154681.000000  154681.000000  154681.000000  154681.000000   
mean     2020.116918    3375.110181     288.751042     193.852025   
std         1.407244    5905.216715     618.369546     430.685453   
min      2018.000000       6.620000   -5172.870000   -4401.040000   
25%      2019.000000     294.480000       9.800000       0.000000   
50%      2020.000000     889.630000      56.060000      33.940000   
75%      2021.000000    3143.960000     237.930000     155.150000   
max      2022.000000   75818.300000   11986.290000    8140.540000   

       other_current_assets  total_current_assets      ppe_gross  \
count         154681.000000         154681.000000  154681.000000   
mean              87.823012           3948.926388    1107.418765   
std              138.114872           6566.452411    1664.659386   
min                1.460000             20.650000      24.880000   
25%               10.340000            379.660000     131.800000   
50%               28.870000           1148.080000     418.190000   
75%               77.650000           3590.700000     925.780000   
max             1438.970000          79587.740000   15722.330000   

       accum_depreciation        ppe_net  other_noncurrent_assets  \
count       154681.000000  154681.000000            154681.000000   
mean           398.552671     708.866085                54.732450   
std            718.549384    1060.421950                88.018025   
min              3.050000      -1.840000                 0.810000   
25%             42.810000      85.000000                 6.420000   
50%            102.360000     271.100000                16.790000   
75%            367.410000     587.570000                50.040000   
max           9304.930000    9374.220000               971.050000   

        total_assets       payables  other_current_liabilities   current_debt  \
count  154681.000000  154681.000000              154681.000000  154681.000000   
mean     4694.271573     273.193630                 153.455840     186.105226   
std      7576.259617     456.466357                 237.774736     390.457097   
min    -39448.690000       1.080000                   2.760000       0.060000   
25%       474.640000      31.140000                  18.090000       9.410000   
50%      1474.580000      79.660000                  52.760000      37.920000   
75%      4157.080000     248.210000                 133.630000     150.910000   
max     85284.540000    5443.020000                2488.900000    6584.190000   

       total_current_liabilities  long_term_debt  total_liabilities  \
count              154681.000000   154681.000000      154681.000000   
mean                  612.754672      813.063519        1425.818189   
std                   983.322038     1401.132337        2323.233216   
min                     9.240000        0.400000          14.680000   
25%                    61.970000       49.930000         113.700000   
50%                   198.380000      188.350000         424.970000   
75%                   506.800000      619.330000        1086.930000   
max                 12750.840000    16861.770000       24978.630000   

        equity_begin      dividends  equity_injection     equity_end  \
count  154681.000000  154681.000000     154681.000000  154681.000000   
mean     2964.576555     168.874954         24.067335    3281.130809   
std      4873.292205     472.890535        138.608120    5626.932181   
min     -5205.810000       0.000000       -135.420000   -5962.450000   
25%       321.240000       0.000000          0.000000     339.260000   
50%       943.620000      16.080000          0.000000     954.170000   
75%      2752.300000     105.580000          0.000000    3125.080000   
max     66006.770000   13332.640000       4000.380000   74964.040000   

       total_liabilities_and_equity  
count                 154681.000000  
mean                    4706.948996  
std                     75

#### Negative value detection
Variable that impossible/not make sense have negative value

*   total_assets = 308 row (after drop recievables and inventory who < 0)
*   recievables = 363 row
*   inventory = 472 row
*   total = 1144(0,7%)
*   total row firm invalid = 3953 (2,5%)

**total negative value recovered = 472 row**
**<br> total row firm recovered = 1482 row (0,9%)**

In [ ]:
balance_sheet_filter2[balance_sheet_filter2['receivables'] < 0]

firm_id  year      cash  receivables  inventory  other_current_assets  \
313     F000069  2021   2568.93      -350.06     419.71                 48.32   
525     F000119  2020  10442.44     -2056.95     740.59                263.64   
685     F000155  2018   1537.19      -367.21       0.00                 92.01   
1203    F000273  2018    191.41       -15.44      25.74                  6.54   
1481    F000332  2021    356.88       -66.22      34.12                 13.88   
...         ...   ...       ...          ...        ...                   ...   
153208  F034109  2020   1866.35      -303.30     101.53                 43.08   
153643  F034214  2020   5234.99      -777.09     479.38                234.92   
153762  F034240  2019    848.97       -13.34      72.91                 37.86   
153901  F034274  2019  10405.12      -641.16     873.48                240.30   
154639  F034440  2020    124.27       -28.88      28.30                  7.71   

        total_current_assets  ppe_gross  accum_depreciation  ppe_net  \
313                  3387.02     718.16              271.14   447.02   
525                 13503.63    3542.25             1179.69  2362.56   
685                  1996.41     766.50              117.30   649.20   
1203                  239.13      99.12               10.19    88.93   
1481                  471.09     174.95               84.38    90.57   
...                      ...        ...                 ...      ...   
153208               2314.27     507.74              167.22   340.53   
153643               6726.38    2181.07              637.22  1543.85   
153762                973.09     328.38               89.97   238.41   
153901              12160.06    3532.48              599.63  2932.85   
154639                189.16      54.29               21.24    33.05   

        other_noncurrent_assets  total_assets  payables  \
313                       54.38       3888.43    293.59   
525                      219.39      16085.57   1197.79   
685                       56.18       2701.80    160.76   
1203                       9.25        337.31     36.73   
1481                       7.36        569.02     19.12   
...                         ...           ...       ...   
153208                    47.16       2701.95    100.48   
153643                    48.38       8318.62    385.44   
153762                    17.67       1229.17     57.72   
153901                   307.88      15400.79   1159.10   
154639                     2.42        224.62     19.05   

        other_current_liabilities  current_debt  total_current_liabilities  \
313                        135.68         78.25                     507.51   
525                        331.22        485.64                    2014.65   
685                        117.72         94.93                     373.40   
1203                        20.49         12.14                      69.36   
1481                        12.36          8.90                      40.38   
...                           ...           ...                        ...   
153208                      77.91        106.88                     285.27   
153643                     389.08        196.58                     971.10   
153762                      50.79         44.21                     152.72   
153901                     894.46        741.37                    2794.92   
154639                       6.88          2.15                      28.08   

        long_term_debt  total_liabilities  equity_begin  dividends  \
313             521.65            1029.16       2782.60      42.43   
525            3237.60            5252.25      11112.64       0.00   
685             176.29             549.69       1655.08     255.27   
1203             22.55              91.91        279.42       0.00   
1481             59.33              99.70        347.11      71.76   
...                ...                ...           ...        ...   
153208          712.56            

In [ ]:
invalid = balance_sheet_filter2[balance_sheet_filter2['receivables'] < 0]

firms_invalid3 = income_valid[income_valid['firm_id'].isin(invalid['firm_id'])]
firms_invalid3.shape

(1726, 13)

In [ ]:
balance_sheet_filter2[balance_sheet_filter2['inventory'] < 0]

firm_id  year     cash  receivables  inventory  other_current_assets  \
173     F000039  2018   552.70         0.00     -44.45                 26.47   
483     F000109  2020  2578.65         3.27     -16.37                 20.13   
743     F000167  2018  1470.96       219.37    -226.09                 54.25   
1558    F000350  2020   251.96        69.59     -40.74                 15.13   
1574    F000353  2021   918.32        17.06      -5.38                 15.00   
...         ...   ...      ...          ...        ...                   ...   
153512  F034180  2021   643.35         0.00      -9.41                  6.18   
153637  F034212  2022  2484.93       184.55     -56.24                 59.89   
153869  F034266  2018  2007.85         0.00    -260.25                 88.29   
153991  F034294  2021   499.37         0.17      -3.79                 12.00   
154679  F034449  2021  2716.84        11.44    -160.35                101.88   

        total_current_assets  ppe_gross  accum_depreciation  ppe_net  \
173                   623.62     139.73               19.17   120.55   
483                  2618.42     498.66              177.58   321.08   
743                  1970.67     445.33               43.70   401.63   
1558                  377.43     219.39               71.64   147.75   
1574                  955.75     217.80              120.06    97.74   
...                      ...        ...                 ...      ...   
153512                658.95     129.14               59.33    69.81   
153637               2785.60    1365.93              293.16  1072.77   
153869               2356.39     697.25               87.28   609.97   
153991                515.33     157.59               83.45    74.14   
154679               2990.51    1227.92              795.41   432.51   

        other_noncurrent_assets  total_assets  payables  \
173                        8.19        752.36     86.36   
483                       27.94       2967.45     55.10   
743                       19.24       2391.54    242.19   
1558                       6.10        531.28     91.65   
1574                      14.28       1067.77     46.66   
...                         ...           ...       ...   
153512                     8.16        736.92     20.30   
153637                    97.61       3955.97    273.51   
153869                    57.89       3024.25    184.50   
153991                     3.95        593.42     39.61   
154679                    23.35       3446.37    360.63   

        other_current_liabilities  current_debt  total_current_liabilities  \
173                         22.99         29.75                     139.10   
483                         53.60         93.18                     201.88   
743                        119.83        318.41                     680.42   
1558                        27.79         12.26                     131.70   
1574                        18.53         25.01                      90.20   
...                           ...           ...                        ...   
153512                      18.59          8.65                      47.54   
153637                     104.06         94.14                     471.71   
153869                     126.77        230.91                     542.18   
153991                      18.79         10.23                      68.64   
154679                     190.51        133.61                     684.75   

        long_term_debt  total_liabilities  equity_begin  dividends  \
173              55.25             194.35        530.83       0.23   
483             621.19             823.06       1948.92      58.12   
743             591.33            1271.75       1223.80       0.00   
1558             81.75             213.45        459.82       0.00   
1574            166.75             256.95        743.06     101.64   
...                ...                ...           ...        ...   
153512           57.69             105.23     

In [ ]:
invalid = balance_sheet_filter2[balance_sheet_filter2['inventory'] < 0]

firms_invalid3 = income_valid[income_valid['firm_id'].isin(invalid['firm_id'])]
firms_invalid3.shape

(2227, 13)

`receivables` and `inventory` calculations involve variables that are not available in the data, namely `days_receivables` and `days_inventory`. so for negative values in these variables will be dropped. but the data will be stored in the `invalid_data` folder.

In [ ]:
balance_sheet_valid = balance_sheet_filter2.copy()

In [ ]:
balance_sheet_receivables_invalid = balance_sheet_valid[balance_sheet_valid['receivables'] < 0]
balance_sheet_receivables_invalid.to_csv('/content/invalid_data/balance_sheet_receivables_invalid.csv')

In [ ]:
balance_sheet_inventory_invalid = balance_sheet_valid[balance_sheet_valid['inventory'] < 0]
balance_sheet_inventory_invalid.to_csv('/content/invalid_data/balance_sheet_inventory_invalid.csv')

In [ ]:
balance_sheet_valid2 = balance_sheet_valid[~((balance_sheet_valid['inventory'] < 0) | (balance_sheet_valid['receivables'] < 0))]
balance_sheet_valid2.shape

(153847, 23)

In [ ]:
balance_sheet_valid2[balance_sheet_valid2['total_assets'] < 0]

firm_id  year      cash  receivables  inventory  other_current_assets  \
1815    F000409  2020  16609.53      1523.38       0.00                419.34   
1866    F000421  2022   5305.15       311.25       0.00                109.70   
2791    F000626  2020    512.91        16.30       0.00                  4.62   
2995    F000671  2020   6322.73       250.84     102.48                165.23   
3342    F000748  2018    301.36         0.00       3.51                  8.81   
...         ...   ...       ...          ...        ...                   ...   
149160  F033185  2020    182.21        16.80      16.69                  4.81   
151090  F033629  2020   3367.03       210.42       0.00                 34.25   
152941  F034050  2019   9180.13       187.03     935.66                186.32   
153597  F034203  2018   2739.86        86.17      50.46                 80.01   
154711  F034457  2018   1858.79       266.27     195.87                 45.64   

        total_current_assets  ppe_gross  accum_depreciation  ppe_net  \
1815                18552.24    4832.24             1548.49  3283.75   
1866                 5726.10     877.24              486.80   390.44   
2791                  533.83      80.99               43.02    37.97   
2995                 6841.27    2901.51             1170.57  1730.94   
3342                  313.67      64.49                9.46    55.02   
...                      ...        ...                 ...      ...   
149160                220.51     112.11               31.94    80.17   
151090               3611.70     688.25              340.62   347.63   
152941              10489.14    2648.78              630.97  2017.81   
153597               2956.51     859.53               98.77   760.77   
154711               2366.57     517.03               86.72   430.30   

        other_noncurrent_assets  total_assets  payables  \
1815                     176.48     -22012.47    899.13   
1866                      58.78      -6175.32    149.10   
2791                       2.27       -574.06     19.25   
2995                     136.28      -8708.50    798.59   
3342                       5.74       -374.44     11.09   
...                         ...           ...       ...   
149160                     1.53       -302.21      8.71   
151090                    25.69      -3985.01     41.88   
152941                   116.32     -12623.27   1562.53   
153597                    69.35      -3786.63    187.63   
154711                    24.80      -2821.68    115.39   

        other_current_liabilities  current_debt  total_current_liabilities  \
1815                       638.77        687.15                    2225.05   
1866                       102.65         53.62                     305.37   
2791                         8.70          7.66                      35.62   
2995                       220.69        505.98                    1525.25   
3342                         8.54         30.56                      50.19   
...                           ...           ...                        ...   
149160                       7.20          4.15                      20.06   
151090                      93.10         61.99                     196.97   
152941                     601.40        411.93                    2575.86   
153597                      99.05        283.91                     570.59   
154711                      71.60        178.12                     365.11   

        long_term_debt  total_liabilities  equity_begin  dividends  \
1815           4581.02            6806.06      14291.33    1372.62   
1866            357.49             662.86       4556.66     806.50   
2791             51.09              86.71        397.02      16.45   
2995           3373.18            4898.43       3603.21     163.09   
3342             56.76             106.95        183.29      17.56   
...                ...                ...           ...        ...   
149160           27.65            

There is negative value from `total_assets` variable, which is not possible. that variable come from this calculation
<br> `total_current_assets` + `ppe_net` + `other_noncurrent_assets`

<br> Negative value from `total_assets` will be convert to NaN. and will be replaced with the formula above

In [ ]:
invalid = balance_sheet_valid2[balance_sheet_valid2['total_assets'] < 0]

firms_invalid3 = income_valid[income_valid['firm_id'].isin(invalid['firm_id'])]
firms_invalid3.shape

(1482, 13)

In [ ]:
balance_sheet_valid2['total_assets'] = np.where(
    balance_sheet_valid2['total_assets'] < 0,
    np.nan,
    balance_sheet_valid2['total_assets']
)
balance_sheet_valid2.isnull().sum()

/tmp/ipython-input-2150245407.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  balance_sheet_valid2['total_assets'] = np.where(


firm_id                           0
year                              0
cash                              0
receivables                       0
inventory                         0
other_current_assets              0
total_current_assets              0
ppe_gross                         0
accum_depreciation                0
ppe_net                           0
other_noncurrent_assets           0
total_assets                    308
payables                          0
other_current_liabilities         0
current_debt                      0
total_current_liabilities         0
long_term_debt                    0
total_liabilities                 0
equity_begin                      0
dividends                         0
equity_injection                  0
equity_end                        0
total_liabilities_and_equity      0
dtype: int64

In [ ]:
balance_sheet_valid2['total_assets'].fillna(balance_sheet_valid2['total_current_assets'] + balance_sheet_valid2['ppe_net'] + balance_sheet_valid2['other_noncurrent_assets'], inplace=True)
balance_sheet_valid2.isnull().sum()

/tmp/ipython-input-2833540021.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  balance_sheet_valid2['total_assets'].fillna(balance_sheet_valid2['total_current_assets'] + balance_sheet_valid2['ppe_net'] + balance_sheet_valid2['other_noncurrent_assets'], inplace=True)
/tmp/ipython-input-2833540021.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-

firm_id                         0
year                            0
cash                            0
receivables                     0
inventory                       0
other_current_assets            0
total_current_assets            0
ppe_gross                       0
accum_depreciation              0
ppe_net                         0
other_noncurrent_assets         0
total_assets                    0
payables                        0
other_current_liabilities       0
current_debt                    0
total_current_liabilities       0
long_term_debt                  0
total_liabilities               0
equity_begin                    0
dividends                       0
equity_injection                0
equity_end                      0
total_liabilities_and_equity    0
dtype: int64

In [ ]:
balance_sheet_valid2.shape

(153847, 23)

### **Cash Flow Statement**
*   There is no NULL value
*   Impossible value `total_assets` < 0 = 3491 (2,2%)

**total recovered = 2.2%**

In [ ]:
cash_flow.head()

firm_id  year  net_income  depreciation  change_receivables  \
0  F000001  2021     3657.82        609.00             1723.84   
1  F000001  2022     4455.57        473.09              975.62   
2  F000002  2018      -67.37         59.04              260.10   
3  F000002  2019      267.97         62.15             -202.55   
4  F000002  2020      504.51         87.93               31.53   

   change_inventory  change_payables  cash_flow_operations    capex  \
0              0.00           963.81               3506.80  4300.20   
1              0.00           -26.92               3926.12   262.82   
2            272.04           193.44               -347.03   420.47   
3            -87.11            53.23                673.01    63.32   
4            -66.91          -178.15                449.67    81.59   

   asset_disposal_proceeds  cash_flow_investing  change_long_term_debt  \
0                      0.0             -4300.20                4939.78   
1                      0.0              -262.82                -229.25   
2                      0.0              -420.47                 171.57   
3                      0.0               -63.32                  -8.65   
4                      0.0               -81.59                   0.00   

   change_current_debt  equity_injection  dividends_paid  cash_flow_financing  \
0              2659.88              0.00          972.29              6627.38   
1             -1953.30              0.00         1215.52             -3398.07   
2                92.38             69.63            0.00               333.58   
3               -67.95              0.00          117.01              -193.61   
4                 0.00              0.00           31.45               -31.45   

   net_cash_flow  cash_beginning  cash_ending  
0        5833.98            0.00      5833.98  
1         265.23        16122.08     16387.31  
2        -433.92            0.00      -433.92  
3         416.08          948.86      1364.94  
4         336.63         1357.06      1693.69

In [ ]:
cash_flow.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 154901 entries, 0 to 154900
Data columns (total 19 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   firm_id                  154901 non-null  object 
 1   year                     154901 non-null  int64  
 2   net_income               154901 non-null  float64
 3   depreciation             154901 non-null  float64
 4   change_receivables       154901 non-null  float64
 5   change_inventory         154901 non-null  float64
 6   change_payables          154901 non-null  float64
 7   cash_flow_operations     154901 non-null  float64
 8   capex                    154901 non-null  float64
 9   asset_disposal_proceeds  154901 non-null  float64
 10  cash_flow_investing      154901 non-null  float64
 11  change_long_term_debt    154901 non-null  float64
 12  change_current_debt      154901 non-null  float64
 13  equity_injection         154901 non-null  float64
 14  divi

#### Filter invalid firm from previous cleaning
Because the problematic data is focused on a certain year and a certain firm too. I decided to delete all firm data that has invalid data problems that cause it to be dropped. because it will be fatal in the next calculation.

In [ ]:
filter3 = pd.read_csv('/content/invalid_data/balance_sheet_inventory_invalid.csv')
filter3['firm_id'].unique()

array(['F000039', 'F000109', 'F000167', 'F000350', 'F000353', 'F000422',
       'F000428', 'F000493', 'F000599', 'F000791', 'F000995', 'F001045',
       'F001134', 'F001153', 'F001165', 'F001224', 'F001356', 'F001381',
       'F001389', 'F001431', 'F001437', 'F001519', 'F001666', 'F001741',
       'F001783', 'F001820', 'F001864', 'F001942', 'F001953', 'F002005',
       'F002054', 'F002447', 'F002461', 'F002484', 'F002860', 'F002920',
       'F002939', 'F002985', 'F003002', 'F003192', 'F003275', 'F003277',
       'F003437', 'F003516', 'F003570', 'F003576', 'F003618', 'F003623',
       'F003643', 'F003711', 'F003734', 'F003799', 'F003826', 'F003867',
       'F003938', 'F003941', 'F003955', 'F004043', 'F004315', 'F004419',
       'F004458', 'F004500', 'F004510', 'F004537', 'F004658', 'F004683',
       'F004809', 'F004930', 'F004939', 'F005251', 'F005351', 'F005389',
       'F005422', 'F005582', 'F005628', 'F005919', 'F005996', 'F006092',
       'F006221', 'F006258', 'F006272', 'F006457', 

In [ ]:
all_invalid_firm_ids = pd.concat([pd.Series(invalid_firm_ids), filter3['firm_id']]).unique()
print(f'Total number of all invalid firm IDs: {len(all_invalid_firm_ids)}')
print(f'Example invalid firm IDs: {all_invalid_firm_ids[:5]}')

Total number of all invalid firm IDs: 512
Example invalid firm IDs: ['F000212' 'F000217' 'F002732' 'F002876' 'F004303']


In [ ]:
filter4 = pd.read_csv('/content/invalid_data/balance_sheet_receivables_invalid.csv')
filter4['firm_id'].unique()

array(['F000069', 'F000119', 'F000155', 'F000273', 'F000332', 'F000340',
       'F000353', 'F000354', 'F000504', 'F000590', 'F000593', 'F000813',
       'F001001', 'F001037', 'F001290', 'F001456', 'F001704', 'F001832',
       'F001886', 'F002000', 'F002105', 'F002152', 'F002206', 'F002257',
       'F002300', 'F002334', 'F002442', 'F002661', 'F002674', 'F002722',
       'F002733', 'F002789', 'F002875', 'F002997', 'F003090', 'F003251',
       'F003257', 'F003322', 'F003331', 'F003438', 'F003668', 'F003725',
       'F003735', 'F003804', 'F003906', 'F004351', 'F004419', 'F004431',
       'F004530', 'F004677', 'F004825', 'F004836', 'F004925', 'F005063',
       'F005145', 'F005159', 'F005292', 'F005557', 'F005579', 'F005778',
       'F005852', 'F005902', 'F005934', 'F005991', 'F006006', 'F006239',
       'F006270', 'F006300', 'F006533', 'F006816', 'F006965', 'F006981',
       'F007030', 'F007044', 'F007116', 'F007166', 'F007177', 'F007191',
       'F007250', 'F007316', 'F007377', 'F007490', 

In [ ]:
all_invalid_firm_ids = pd.concat([pd.Series(all_invalid_firm_ids), filter4['firm_id']]).unique()
print(f'Total number of all invalid firm IDs: {len(all_invalid_firm_ids)}')
print(f'Example invalid firm IDs: {all_invalid_firm_ids[:5]}')

Total number of all invalid firm IDs: 870
Example invalid firm IDs: ['F000212' 'F000217' 'F002732' 'F002876' 'F004303']


In [ ]:
all_invalid_firm_ids

array(['F000212', 'F000217', 'F002732', 'F002876', 'F004303', 'F006122',
       'F006804', 'F007461', 'F008189', 'F008781', 'F010425', 'F010863',
       'F011925', 'F012700', 'F013947', 'F014531', 'F015133', 'F015944',
       'F016008', 'F016027', 'F016309', 'F017404', 'F018015', 'F018220',
       'F018257', 'F019407', 'F021619', 'F021985', 'F022771', 'F024114',
       'F024294', 'F025963', 'F026459', 'F027102', 'F027143', 'F027203',
       'F029762', 'F031930', 'F032049', 'F033046', 'F033091', 'F033329',
       'F033397', 'F018176', 'F025025', 'F000039', 'F000109', 'F000167',
       'F000350', 'F000353', 'F000422', 'F000428', 'F000493', 'F000599',
       'F000791', 'F000995', 'F001045', 'F001134', 'F001153', 'F001165',
       'F001224', 'F001356', 'F001381', 'F001389', 'F001431', 'F001437',
       'F001519', 'F001666', 'F001741', 'F001783', 'F001820', 'F001864',
       'F001942', 'F001953', 'F002005', 'F002054', 'F002447', 'F002461',
       'F002484', 'F002860', 'F002920', 'F002939', 

In [ ]:
cash_flow_filter = cash_flow[~cash_flow['firm_id'].isin(all_invalid_firm_ids)]
cash_flow_filter[cash_flow_filter['firm_id'] == 'F034440']

Empty DataFrame
Columns: [firm_id, year, net_income, depreciation, change_receivables, change_inventory, change_payables, cash_flow_operations, capex, asset_disposal_proceeds, cash_flow_investing, change_long_term_debt, change_current_debt, equity_injection, dividends_paid, cash_flow_financing, net_cash_flow, cash_beginning, cash_ending]
Index: []

In [ ]:
income_filter = income_valid[~income_valid['firm_id'].isin(all_invalid_firm_ids)]
income_filter[income_filter['firm_id'] == 'F034109']

Empty DataFrame
Columns: [firm_id, year, gross_profit, ebitda, depreciation, interest_expense, ebt, tax, net_income, revenue_valid, cogs_valid, opex_valid, ebit_valid]
Index: []

In [ ]:
balance_sheet_final = balance_sheet_valid2[~balance_sheet_valid2['firm_id'].isin(all_invalid_firm_ids)]
balance_sheet_final[balance_sheet_final['firm_id'] == 'F034440']

Empty DataFrame
Columns: [firm_id, year, cash, receivables, inventory, other_current_assets, total_current_assets, ppe_gross, accum_depreciation, ppe_net, other_noncurrent_assets, total_assets, payables, other_current_liabilities, current_debt, total_current_liabilities, long_term_debt, total_liabilities, equity_begin, dividends, equity_injection, equity_end, total_liabilities_and_equity]
Index: []

In [ ]:
company_info_filter = company_info.copy()

In [ ]:
company_info_filter = company_info_filter[~company_info_filter['firm_id'].isin(all_invalid_firm_ids)]
company_info_filter[company_info_filter['firm_id'] == 'F034440']

Empty DataFrame
Columns: [firm_id, sector, region, start_year]
Index: []

In [ ]:
income_filter.shape

(150748, 13)

In [ ]:
income_clean = income_filter.copy()
income_clean.to_csv('/content/clean_data/income_clean.csv', index=False)

In [ ]:
cash_flow_filter.shape

(150748, 19)

In [ ]:
cash_flow_clean = cash_flow_filter.copy()
cash_flow_clean.to_csv('/content/clean_data/cash_flow_clean.csv', index=False)

In [ ]:
balance_sheet_final.shape

(150748, 23)

In [ ]:
balance_sheet_clean = balance_sheet_final.copy()
balance_sheet_clean.to_csv('/content/clean_data/balance_sheet_clean.csv', index=False)

In [ ]:
company_info_filter.shape

(33630, 4)

#### impossible Value
*   there is value `net_income` in cash flow statement which is different from `income_statement['net_income']` which should be the same value.

In [ ]:
cash_flow_valid = cash_flow_filter.copy()

Create a new variable named `net_income_diff` and calculate the difference between the two variables.

In [ ]:
cash_flow_valid['net_income_diff'] = cash_flow_valid['net_income'] - income_filter['net_income']
cash_flow_valid.head()

firm_id  year  net_income  depreciation  change_receivables  \
0  F000001  2021     3657.82        609.00             1723.84   
1  F000001  2022     4455.57        473.09              975.62   
2  F000002  2018      -67.37         59.04              260.10   
3  F000002  2019      267.97         62.15             -202.55   
4  F000002  2020      504.51         87.93               31.53   

   change_inventory  change_payables  cash_flow_operations    capex  \
0              0.00           963.81               3506.80  4300.20   
1              0.00           -26.92               3926.12   262.82   
2            272.04           193.44               -347.03   420.47   
3            -87.11            53.23                673.01    63.32   
4            -66.91          -178.15                449.67    81.59   

   asset_disposal_proceeds  cash_flow_investing  change_long_term_debt  \
0                      0.0             -4300.20                4939.78   
1                      0.0              -262.82                -229.25   
2                      0.0              -420.47                 171.57   
3                      0.0               -63.32                  -8.65   
4                      0.0               -81.59                   0.00   

   change_current_debt  equity_injection  dividends_paid  cash_flow_financing  \
0              2659.88              0.00          972.29              6627.38   
1             -1953.30              0.00         1215.52             -3398.07   
2                92.38             69.63            0.00               333.58   
3               -67.95              0.00          117.01              -193.61   
4                 0.00              0.00           31.45               -31.45   

   net_cash_flow  cash_beginning  cash_ending  net_income_diff  
0        5833.98            0.00      5833.98              0.0  
1         265.23        16122.08     16387.31              0.0  
2        -433.92            0.00      -433.92              0.0  
3         416.08          948.86      1364.94              0.0  
4         336.63         1357.06      1693.69              0.0

In [ ]:
cash_flow_valid[cash_flow_valid['net_income_diff']>0]

firm_id  year   net_income  depreciation  change_receivables  \
12      F000004  2019     2.994849         10.06                0.00   
382     F000084  2022  2706.135617        323.23                0.00   
530     F000120  2020    26.485039        151.94              119.23   
922     F000207  2020   144.308325        611.72              306.75   
1052    F000237  2021   121.207253         23.79              -16.64   
...         ...   ...          ...           ...                 ...   
153816  F034255  2019   149.968916         19.12               10.26   
153894  F034271  2019  1274.105239        131.88               60.24   
154265  F034356  2022    14.961954         29.97              313.64   
154414  F034392  2020     4.372624         11.15                0.00   
154713  F034457  2020   772.104519         64.73              139.75   

        change_inventory  change_payables  cash_flow_operations   capex  \
12                 39.63            33.26                -18.46   99.30   
382               103.75            43.22               1495.93  409.69   
530               -52.17           -70.14               -197.71   34.98   
922              1264.39           183.38              -1959.39  225.22   
1052               -1.16             2.42                101.46   11.57   
...                  ...              ...                   ...     ...   
153816              0.00            40.60                120.87   20.11   
153894              0.00            40.02                710.40   56.44   
154265            186.62            99.03               -440.55  310.90   
154414             -6.86           -35.50                -50.59    3.93   
154713             -8.33          -111.12                177.16   68.71   

        asset_disposal_proceeds  cash_flow_investing  change_long_term_debt  \
12                          0.0               -99.30                  93.75   
382                         0.0              -409.69                -206.11   
530                         0.0               -34.98                   0.00   
922                         0.0              -225.22                 450.26   
1052                        0.0               -11.57                   5.40   
...                         ...                  ...                    ...   
153816                      0.0               -20.11                   0.20   
153894                      0.0               -56.44                  -8.65   
154265                      0.0              -310.90                 330.06   
154414                      0.0                -3.93                   0.00   
154713                      0.0               -68.71                  64.58   

        change_current_debt  equity_injection  dividends_paid  \
12                    50.48              0.00            0.00   
382                  -30.92              0.00          347.46   
530                    0.00              0.00            0.00   
922                   67.54              0.00            0.00   
1052                   0.81              0.00           18.18   
...                     ...               ...             ...   
153816               -14.09              0.00            6.12   
153894              -136.85              0.00          187.54   
154265               177.72              0.00            0.00   
154414                 0.00              0.00            0.00   
154713                 9.69            224.91          152.56   

        cash_flow_financing  net_cash_flow  cash_beginning  cash_ending  \
12                   144.23          26.47            0.00        26.47   
382                 -584.49         501.74        10060.52     10562.26   
530                    0.00        -232.69          521.97       289.28   
922                  517.80       -1666.81        11868.61     10201.80   
1052                 -11.98          77.91          431.81       509.73   
...                     ...            ...             ...    

In [ ]:
invalid = cash_flow_valid[cash_flow_valid['net_income_diff']>0]

firms_invalid3 = cash_flow_valid[cash_flow_valid['firm_id'].isin(invalid['firm_id'])]
firms_invalid3.shape

(3491, 20)

In [ ]:
cash_flow_valid.isnull().sum()

firm_id                    0
year                       0
net_income                 0
depreciation               0
change_receivables         0
change_inventory           0
change_payables            0
cash_flow_operations       0
capex                      0
asset_disposal_proceeds    0
cash_flow_investing        0
change_long_term_debt      0
change_current_debt        0
equity_injection           0
dividends_paid             0
cash_flow_financing        0
net_cash_flow              0
cash_beginning             0
cash_ending                0
net_income_diff            0
dtype: int64

variables whose `net_income_diff` value is greater than 0. will be completely replaced with the value in `income_statement['net_income']` because the other values of the variable are patterned after the income statement dataframe.

In [ ]:
income_map = income_filter.set_index(['firm_id', 'year'])['net_income']

mask = cash_flow_valid['net_income_diff'] > 0

# ambil key untuk mapping
keys = list(zip(cash_flow_valid.loc[mask, 'firm_id'], cash_flow_valid.loc[mask, 'year']))

cash_flow_valid.loc[mask, 'net_income'] = [income_map.get(k, v) for k, v in zip(keys, cash_flow_valid.loc[mask, 'net_income'])]

In [ ]:
cash_flow_valid['net_income_diff'] = cash_flow_valid['net_income'] - income_filter['net_income']
cash_flow_valid.head()

firm_id  year  net_income  depreciation  change_receivables  \
0  F000001  2021     3657.82        609.00             1723.84   
1  F000001  2022     4455.57        473.09              975.62   
2  F000002  2018      -67.37         59.04              260.10   
3  F000002  2019      267.97         62.15             -202.55   
4  F000002  2020      504.51         87.93               31.53   

   change_inventory  change_payables  cash_flow_operations    capex  \
0              0.00           963.81               3506.80  4300.20   
1              0.00           -26.92               3926.12   262.82   
2            272.04           193.44               -347.03   420.47   
3            -87.11            53.23                673.01    63.32   
4            -66.91          -178.15                449.67    81.59   

   asset_disposal_proceeds  cash_flow_investing  change_long_term_debt  \
0                      0.0             -4300.20                4939.78   
1                      0.0              -262.82                -229.25   
2                      0.0              -420.47                 171.57   
3                      0.0               -63.32                  -8.65   
4                      0.0               -81.59                   0.00   

   change_current_debt  equity_injection  dividends_paid  cash_flow_financing  \
0              2659.88              0.00          972.29              6627.38   
1             -1953.30              0.00         1215.52             -3398.07   
2                92.38             69.63            0.00               333.58   
3               -67.95              0.00          117.01              -193.61   
4                 0.00              0.00           31.45               -31.45   

   net_cash_flow  cash_beginning  cash_ending  net_income_diff  
0        5833.98            0.00      5833.98              0.0  
1         265.23        16122.08     16387.31              0.0  
2        -433.92            0.00      -433.92              0.0  
3         416.08          948.86      1364.94              0.0  
4         336.63         1357.06      1693.69              0.0

In [ ]:
cash_flow_valid[cash_flow_valid['net_income_diff']>0]

Empty DataFrame
Columns: [firm_id, year, net_income, depreciation, change_receivables, change_inventory, change_payables, cash_flow_operations, capex, asset_disposal_proceeds, cash_flow_investing, change_long_term_debt, change_current_debt, equity_injection, dividends_paid, cash_flow_financing, net_cash_flow, cash_beginning, cash_ending, net_income_diff]
Index: []

##### Start_year > earliest year
Comparing and validating `company_info['start_year']` using `income['year']`

In [ ]:
company_info_valid = company_info_filter.copy()

In [ ]:
earliest_year_df = income.groupby('firm_id', as_index=False)['year'].min().rename(columns={'year': 'earliest_year'})

In [ ]:
company_info_valid = company_info_valid.merge(earliest_year_df, on='firm_id', how='left')

In [ ]:
mask_invalid_start = company_info_valid['start_year'] > company_info_valid['earliest_year']

In [ ]:
invalid_start_firms = company_info_valid.loc[mask_invalid_start, ['firm_id', 'start_year', 'earliest_year']]
print(f"Jumlah perusahaan dengan START_YEAR > EARLIEST YEAR: {mask_invalid_start.sum()}")
display(invalid_start_firms.head(10))

Jumlah perusahaan dengan START_YEAR > EARLIEST YEAR: 269


firm_id  start_year  earliest_year
525   F000544        2020           2018
601   F000623        2019           2018
950   F000974        2021           2020
986   F001012        2020           2019
1078  F001106        2024           2022
1147  F001178        2020           2018
1264  F001297        2021           2020
1312  F001345        2019           2018
1676  F001718        2020           2018
1709  F001752        2021           2019

In [ ]:
company_info_valid.loc[mask_invalid_start, 'start_year'] = (
    company_info_valid.loc[mask_invalid_start, 'earliest_year'] - 1
)

In [ ]:
company_info_valid.drop(columns=['earliest_year'], inplace=True)

In [ ]:
company_info_valid.head()

firm_id  sector      region  start_year
0  F000001    Jasa  Kalimantan        2021
1  F000002  Retail        Jawa        2000
2  F000003    Jasa    Sumatera        1995
3  F000004  Retail  Kalimantan        2019
4  F000005  Retail    Sumatera        2012

In [ ]:
company_info_clean = company_info_valid.copy()
company_info_clean.to_csv('/content/clean_data/company_info_clean.csv', index=False)